File: crates/masonry/src/objects.rs
//! Snapshot descriptions of scenes and game objects.

use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

use crate::{
    CameraClearMode, CameraProjection, Color, FontAddress, HorizontalAlignment, ImageFit,
    LightType, LocalTransform, MaterialAddress, ObjectId, PointerEvent, PrefabAddress, RgbColor,
    SceneAddress, SceneId, ShadowMode, TextureAddress, VerticalAlignment,
};

/// One additively loaded Addressable content-scene instance.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct Scene {
    /// Identity of this scene instance within the session.
    pub scene_id: SceneId,
    /// Prepared Addressables scene address to load.
    pub address: SceneAddress,
}

impl Scene {
    /// Creates a content-scene declaration.
    #[must_use]
    pub fn new(scene_id: SceneId, address: impl Into<SceneAddress>) -> Self {
        Self {
            scene_id,
            address: address.into(),
        }
    }
}

/// A complete game object from a snapshot or `object.create` command.
#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
pub struct GameObject {
    /// Session-unique identity of the game object.
    pub object_id: ObjectId,
    /// Scene that owns the game object.
    pub parent_scene: ParentScene,
    /// Optional parent game object in the same scene.
    pub parent_id: Option<ObjectId>,
    /// The Unity GameObject's `activeSelf` value. Omission means `true`.
    ///
    /// This is the value passed to `GameObject.SetActive`; `activeInHierarchy`
    /// can still be false because of an inactive parent. It is separate from
    /// component `enabled` flags and from Unity's active Scene.
    pub active: bool,
    /// Local transform relative to the parent or placement container.
    pub local_transform: LocalTransform,
    /// Unique pointer events enabled for this object.
    pub pointer_events: Vec<PointerEvent>,
    /// Kind-specific object content and component state.
    pub kind: GameObjectKind,
}

impl GameObject {
    /// Creates a game object in the current primary scene with `activeSelf` set.
    #[must_use]
    pub fn new(object_id: ObjectId, kind: impl Into<GameObjectKind>) -> Self {
        Self {
            object_id,
            parent_scene: ParentScene::PrimaryScene,
            parent_id: None,
            active: true,
            local_transform: LocalTransform::default(),
            pointer_events: Vec::new(),
            kind: kind.into(),
        }
    }
}

/// The scene container that owns a game object.
#[derive(Clone, Copy, Debug, Default, Deserialize, PartialEq, Serialize)]
pub enum ParentScene {
    /// The primary content scene at the time the object is created.
    #[default]
    PrimaryScene,
    /// A specific loaded content-scene instance.
    Scene(SceneId),
    /// Masonry's bootstrap-scene container for objects that survive scene unloads.
    Persistent,
}

/// The concrete content created for a game object.
#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
pub enum GameObjectKind {
    /// An empty GameObject.
    Empty,
    /// Unity's standard cube primitive.
    Cube {
        /// Ordered prepared-material assignments with unique renderer slots.
        materials: Vec<MaterialAssignment>,
    },
    /// Unity's standard sphere primitive.
    Sphere {
        /// Ordered prepared-material assignments with unique renderer slots.
        materials: Vec<MaterialAssignment>,
    },
    /// Unity's standard capsule primitive.
    Capsule {
        /// Ordered prepared-material assignments with unique renderer slots.
        materials: Vec<MaterialAssignment>,
    },
    /// Unity's standard cylinder primitive.
    Cylinder {
        /// Ordered prepared-material assignments with unique renderer slots.
        materials: Vec<MaterialAssignment>,
    },
    /// Unity's standard plane primitive.
    Plane {
        /// Ordered prepared-material assignments with unique renderer slots.
        materials: Vec<MaterialAssignment>,
    },
    /// Unity's standard quad primitive.
    Quad {
        /// Ordered prepared-material assignments with unique renderer slots.
        materials: Vec<MaterialAssignment>,
    },
    /// A Masonry-owned image quad.
    Image {
        /// Complete image component state.
        image: ImageState,
    },
    /// A world-space TextMesh Pro object.
    Text {
        /// Complete text component state.
        text: TextState,
    },
    /// A standard Unity camera.
    Camera {
        /// Complete camera component state.
        camera: CameraState,
    },
    /// A standard Unity light.
    Light {
        /// Complete light component state.
        light: LightState,
    },
    /// An instance of a prepared prefab.
    Prefab {
        /// Prepared prefab address.
        address: PrefabAddress,
        /// Ordered prepared-material assignments with unique renderer slots.
        materials: Vec<MaterialAssignment>,
        /// Stable Animator state, when the prefab has an Animator.
        animator: Option<AnimatorState>,
    },
}

impl GameObjectKind {
    /// Creates a prefab instance without material or Animator overrides.
    #[must_use]
    pub fn prefab(address: impl Into<PrefabAddress>) -> Self {
        Self::Prefab {
            address: address.into(),
            materials: Vec::new(),
            animator: None,
        }
    }

    /// Creates a cube without material overrides.
    #[must_use]
    pub fn cube() -> Self {
        Self::Cube {
            materials: Vec::new(),
        }
    }

    /// Creates a sphere without material overrides.
    #[must_use]
    pub fn sphere() -> Self {
        Self::Sphere {
            materials: Vec::new(),
        }
    }

    /// Creates a capsule without material overrides.
    #[must_use]
    pub fn capsule() -> Self {
        Self::Capsule {
            materials: Vec::new(),
        }
    }

    /// Creates a cylinder without material overrides.
    #[must_use]
    pub fn cylinder() -> Self {
        Self::Cylinder {
            materials: Vec::new(),
        }
    }

    /// Creates a plane without material overrides.
    #[must_use]
    pub fn plane() -> Self {
        Self::Plane {
            materials: Vec::new(),
        }
    }

    /// Creates a quad without material overrides.
    #[must_use]
    pub fn quad() -> Self {
        Self::Quad {
            materials: Vec::new(),
        }
    }
}

impl From<ImageState> for GameObjectKind {
    fn from(image: ImageState) -> Self {
        Self::Image { image }
    }
}

impl From<TextState> for GameObjectKind {
    fn from(text: TextState) -> Self {
        Self::Text { text }
    }
}

impl From<CameraState> for GameObjectKind {
    fn from(camera: CameraState) -> Self {
        Self::Camera { camera }
    }
}

impl From<LightState> for GameObjectKind {
    fn from(light: LightState) -> Self {
        Self::Light { light }
    }
}

/// One prepared material assigned to a prefab renderer slot.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct MaterialAssignment {
    /// Zero-based index in the renderer's shared-material array.
    pub slot: u32,
    /// Prepared material address assigned to the slot.
    pub address: MaterialAddress,
}

impl MaterialAssignment {
    /// Creates one renderer material-slot assignment.
    #[must_use]
    pub fn new(slot: u32, address: impl Into<MaterialAddress>) -> Self {
        Self {
            slot,
            address: address.into(),
        }
    }
}

/// Complete state for a Masonry-owned image quad.
#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
pub struct ImageState {
    /// Prepared texture address.
    pub texture: TextureAddress,
    /// Positive world-space width around a centered pivot.
    pub width: f64,
    /// Positive world-space height around a centered pivot.
    pub height: f64,
    /// How the texture fits the requested dimensions.
    pub fit: ImageFit,
    /// Linear RGB tint; opacity is controlled separately.
    pub tint: RgbColor,
    /// Opacity in the inclusive range `[0, 1]`.
    pub opacity: f64,
    /// Whether the image rotates to face the input camera.
    pub face_camera: bool,
}

impl ImageState {
    /// Creates opaque, untinted image state using stretch fitting.
    #[must_use]
    pub fn new(texture: impl Into<TextureAddress>, width: f64, height: f64) -> Self {
        Self {
            texture: texture.into(),
            width,
            height,
            fit: ImageFit::Stretch,
            tint: RgbColor::WHITE,
            opacity: 1.0,
            face_camera: false,
        }
    }
}

/// Complete state for a world-space TextMesh Pro object.
#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
pub struct TextState {
    /// Displayed text content.
    pub text: String,
    /// Prepared TextMesh Pro font address.
    pub font: FontAddress,
    /// Positive world-space text size.
    pub size: f64,
    /// Linear text color.
    pub color: Color,
    /// Horizontal text alignment.
    pub horizontal: HorizontalAlignment,
    /// Vertical text alignment.
    pub vertical: VerticalAlignment,
    /// Positive wrapping width; [`None`] disables wrapping.
    pub wrap_width: Option<f64>,
    /// Whether TextMesh Pro rich-text tags are interpreted.
    pub rich_text: bool,
    /// Whether the text rotates to face the input camera.
    pub face_camera: bool,
}

impl TextState {
    /// Creates centered, unwrapped, opaque-white world-text state at size one.
    #[must_use]
    pub fn new(text: impl Into<String>, font: impl Into<FontAddress>) -> Self {
        Self {
            text: text.into(),
            font: font.into(),
            size: 1.0,
            color: Color::WHITE,
            horizontal: HorizontalAlignment::Center,
            vertical: VerticalAlignment::Middle,
            wrap_width: None,
            rich_text: false,
            face_camera: false,
        }
    }
}

/// Complete state for a standard Unity camera.
#[derive(Clone, Copy, Debug, Deserialize, PartialEq, Serialize)]
pub struct CameraState {
    /// Whether the Camera component is enabled.
    pub enabled: bool,
    /// Perspective or orthographic projection.
    pub projection: CameraProjection,
    /// Perspective vertical field of view in degrees, strictly between 1 and 179.
    pub field_of_view: f64,
    /// Positive orthographic half-height.
    pub orthographic_size: f64,
    /// Positive near clipping distance.
    pub near: f64,
    /// Far clipping distance, which must be greater than `near`.
    pub far: f64,
    /// Camera clear behavior.
    pub clear_mode: CameraClearMode,
    /// Linear color used by solid-color clearing.
    pub clear_color: Color,
}

impl Default for CameraState {
    fn default() -> Self {
        Self {
            enabled: true,
            projection: CameraProjection::Perspective,
            field_of_view: 60.0,
            orthographic_size: 5.0,
            near: 0.3,
            far: 1000.0,
            clear_mode: CameraClearMode::Skybox,
            clear_color: Color::BLACK,
        }
    }
}

impl CameraState {
    /// Creates camera state with its defaults.
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }
}

/// Complete state for a standard Unity light.
#[derive(Clone, Copy, Debug, Deserialize, PartialEq, Serialize)]
pub struct LightState {
    /// Whether the Light component is enabled.
    pub enabled: bool,
    /// Directional, point, or spot behavior.
    pub light_type: LightType,
    /// Linear light color.
    pub color: Color,
    /// Nonnegative light intensity.
    pub intensity: f64,
    /// Positive range for point and spot lights.
    pub range: f64,
    /// Outer spot angle in degrees, strictly between 0 and 179.
    pub outer_spot_angle: f64,
    /// Inner spot angle in degrees, between zero and `outer_spot_angle`.
    pub inner_spot_angle: f64,
    /// Shadow rendering mode.
    pub shadows: ShadowMode,
}

impl Default for LightState {
    fn default() -> Self {
        Self {
            enabled: true,
            light_type: LightType::Point,
            color: Color::WHITE,
            intensity: 1.0,
            range: 10.0,
            outer_spot_angle: 30.0,
            inner_spot_angle: 0.0,
            shadows: ShadowMode::None,
        }
    }
}

impl LightState {
    /// Creates light state with its defaults.
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }
}

/// Stable Animator state reconstructed by a snapshot.
#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
pub struct AnimatorState {
    /// Animator state name to play.
    pub state: String,
    /// Nonnegative Animator layer index.
    pub layer: u32,
    /// Normalized starting time in the inclusive range `[0, 1]`.
    pub normalized_start_time: f64,
    /// Persistent boolean parameters, ordered by name for deterministic output.
    pub bool_parameters: BTreeMap<String, bool>,
    /// Persistent signed 32-bit integer parameters, ordered by name.
    pub int_parameters: BTreeMap<String, i32>,
    /// Persistent finite floating-point parameters, ordered by name.
    pub float_parameters: BTreeMap<String, f64>,
    /// Nonnegative Animator playback speed.
    pub speed: f64,
}

impl AnimatorState {
    /// Creates stable Animator state on layer zero at normalized time zero.
    #[must_use]
    pub fn new(state: impl Into<String>) -> Self {
        Self {
            state: state.into(),
            layer: 0,
            normalized_start_time: 0.0,
            bool_parameters: BTreeMap::new(),
            int_parameters: BTreeMap::new(),
            float_parameters: BTreeMap::new(),
            speed: 1.0,
        }
    }
}

File: Packages/com.masonry.client/Runtime/Protocol/Objects.cs
#nullable enable

using System;
using System.Collections.Generic;

namespace Masonry
{
    /// <summary>One additively loaded Addressable content-scene instance.</summary>
    /// <param name="Id">Identity of this scene instance within the session.</param>
    /// <param name="Address">Prepared Addressables scene address to load.</param>
    public sealed record MasonryScene(SceneId Id, SceneAddress Address);

    /// <summary>A complete game object from a snapshot or object-create command.</summary>
    /// <param name="Id">Session-unique identity of the game object.</param>
    /// <param name="Kind">Kind-specific object content and component state.</param>
    /// <param name="ParentScene">Scene that owns the game object.</param>
    /// <param name="ParentId">Optional parent game object in the same scene.</param>
    /// <param name="IsActive">The object's activation value.</param>
    /// <param name="LocalTransform">Local transform relative to the parent or placement.</param>
    /// <param name="PointerEvents">Unique pointer events enabled for this object.</param>
    /// <param name="DragMode">Local pointer-following behavior, or null when not draggable.</param>
    public sealed record MasonryGameObject(
        ObjectId Id,
        GameObjectKind Kind,
        ParentScene ParentScene,
        ObjectId? ParentId,
        bool IsActive,
        LocalTransform LocalTransform,
        IReadOnlyList<PointerEvent> PointerEvents,
        DragMode? DragMode = null
    )
    {
        public MasonryGameObject(ObjectId id, GameObjectKind kind)
            : this(
                id,
                kind,
                new ParentScene.Primary(),
                null,
                true,
                LocalTransform.Identity,
                Array.Empty<PointerEvent>(),
                null
            ) { }
    }

    /// <summary>The scene container that owns a game object.</summary>
    public abstract record ParentScene
    {
        private ParentScene() { }

        /// <summary>The primary content scene at the time the object is created.</summary>
        public sealed record Primary : ParentScene;

        /// <summary>A specific loaded content-scene instance.</summary>
        public sealed record Specific(SceneId SceneId) : ParentScene;

        /// <summary>The bootstrap-scene container for objects that survive scene unloads.</summary>
        public sealed record Persistent : ParentScene;
    }

    /// <summary>The concrete content created for a game object.</summary>
    public abstract record GameObjectKind
    {
        private GameObjectKind() { }

        /// <summary>An empty game object.</summary>
        public sealed record Empty : GameObjectKind;

        /// <summary>A standard cube primitive.</summary>
        public sealed record Cube(IReadOnlyList<MaterialAssignment> Materials) : GameObjectKind
        {
            public Cube()
                : this(Array.Empty<MaterialAssignment>()) { }
        }

        /// <summary>A standard sphere primitive.</summary>
        public sealed record Sphere(IReadOnlyList<MaterialAssignment> Materials) : GameObjectKind
        {
            public Sphere()
                : this(Array.Empty<MaterialAssignment>()) { }
        }

        /// <summary>A standard capsule primitive.</summary>
        public sealed record Capsule(IReadOnlyList<MaterialAssignment> Materials) : GameObjectKind
        {
            public Capsule()
                : this(Array.Empty<MaterialAssignment>()) { }
        }

        /// <summary>A standard cylinder primitive.</summary>
        public sealed record Cylinder(IReadOnlyList<MaterialAssignment> Materials) : GameObjectKind
        {
            public Cylinder()
                : this(Array.Empty<MaterialAssignment>()) { }
        }

        /// <summary>A standard plane primitive.</summary>
        public sealed record Plane(IReadOnlyList<MaterialAssignment> Materials) : GameObjectKind
        {
            public Plane()
                : this(Array.Empty<MaterialAssignment>()) { }
        }

        /// <summary>A standard quad primitive.</summary>
        public sealed record Quad(IReadOnlyList<MaterialAssignment> Materials) : GameObjectKind
        {
            public Quad()
                : this(Array.Empty<MaterialAssignment>()) { }
        }

        /// <summary>A Masonry-owned image quad.</summary>
        public sealed record Image(ImageState State) : GameObjectKind;

        /// <summary>A world-space TextMesh Pro object.</summary>
        public sealed record Text(TextState State) : GameObjectKind;

        /// <summary>A standard camera.</summary>
        public sealed record Camera(CameraState State) : GameObjectKind;

        /// <summary>A standard light.</summary>
        public sealed record Light(LightState State) : GameObjectKind;

        /// <summary>An instance of a prepared prefab.</summary>
        /// <param name="Address">Prepared prefab address.</param>
        /// <param name="Materials">Ordered material assignments with unique renderer slots.</param>
        /// <param name="Animator">Stable Animator state, when the prefab has an Animator.</param>
        public sealed record Prefab(
            PrefabAddress Address,
            IReadOnlyList<MaterialAssignment> Materials,
            AnimatorState? Animator = null
        ) : GameObjectKind
        {
            public Prefab(PrefabAddress address)
                : this(address, Array.Empty<MaterialAssignment>()) { }
        }
    }

    /// <summary>One prepared material assigned to a prefab renderer slot.</summary>
    /// <param name="Slot">Zero-based index in the renderer's material array.</param>
    /// <param name="Address">Prepared material address assigned to the slot.</param>
    public sealed record MaterialAssignment(uint Slot, MaterialAddress Address);

    /// <summary>Complete state for a Masonry-owned image quad.</summary>
    /// <param name="Texture">Prepared texture address.</param>
    /// <param name="Width">Positive world-space width around a centered pivot.</param>
    /// <param name="Height">Positive world-space height around a centered pivot.</param>
    /// <param name="Fit">How the texture fits the requested dimensions.</param>
    /// <param name="Tint">Linear RGB tint; opacity is controlled separately.</param>
    /// <param name="Opacity">Opacity in the inclusive range [0, 1].</param>
    /// <param name="FacesCamera">Whether the image rotates to face the input camera.</param>
    public sealed record ImageState(
        TextureAddress Texture,
        double Width,
        double Height,
        ImageFit Fit,
        RgbColor Tint,
        double Opacity,
        bool FacesCamera
    )
    {
        public ImageState(TextureAddress texture, double width, double height)
            : this(texture, width, height, ImageFit.Stretch, RgbColor.White, 1, false) { }
    }

    /// <summary>Complete state for a world-space TextMesh Pro object.</summary>
    /// <param name="Text">Displayed text content.</param>
    /// <param name="Font">Prepared TextMesh Pro font address.</param>
    /// <param name="Size">Positive world-space text size.</param>
    /// <param name="Color">Linear text color.</param>
    /// <param name="HorizontalAlignment">Horizontal text alignment.</param>
    /// <param name="VerticalAlignment">Vertical text alignment.</param>
    /// <param name="WrapWidth">Positive wrapping width; null disables wrapping.</param>
    /// <param name="IsRichText">Whether TextMesh Pro rich-text tags are interpreted.</param>
    /// <param name="FacesCamera">Whether the text rotates to face the input camera.</param>
    public sealed record TextState(
        string Text,
        FontAddress Font,
        double Size,
        Color Color,
        HorizontalAlignment HorizontalAlignment,
        VerticalAlignment VerticalAlignment,
        double? WrapWidth,
        bool IsRichText,
        bool FacesCamera
    )
    {
        public TextState(string text, FontAddress font)
            : this(
                text,
                font,
                1,
                Masonry.Color.White,
                HorizontalAlignment.Center,
                VerticalAlignment.Middle,
                null,
                false,
                false
            ) { }
    }

    /// <summary>Complete state for a standard camera.</summary>
    /// <param name="IsEnabled">Whether the Camera component is enabled.</param>
    /// <param name="Projection">Perspective or orthographic projection.</param>
    /// <param name="FieldOfView">Perspective vertical field of view in degrees.</param>
    /// <param name="OrthographicSize">Positive orthographic half-height.</param>
    /// <param name="NearClip">Positive near clipping distance.</param>
    /// <param name="FarClip">Far clipping distance, greater than the near distance.</param>
    /// <param name="ClearMode">Camera clear behavior.</param>
    /// <param name="ClearColor">Linear color used by solid-color clearing.</param>
    public sealed record CameraState(
        bool IsEnabled,
        CameraProjection Projection,
        double FieldOfView,
        double OrthographicSize,
        double NearClip,
        double FarClip,
        CameraClearMode ClearMode,
        Color ClearColor
    )
    {
        public CameraState()
            : this(
                true,
                CameraProjection.Perspective,
                60,
                5,
                0.3,
                1000,
                CameraClearMode.Skybox,
                Masonry.Color.Black
            ) { }
    }

    /// <summary>Complete state for a standard light.</summary>
    /// <param name="IsEnabled">Whether the Light component is enabled.</param>
    /// <param name="Type">Directional, point, or spot behavior.</param>
    /// <param name="Color">Linear light color.</param>
    /// <param name="Intensity">Nonnegative light intensity.</param>
    /// <param name="Range">Positive range for point and spot lights.</param>
    /// <param name="OuterSpotAngle">Outer spot angle in degrees.</param>
    /// <param name="InnerSpotAngle">Inner spot angle in degrees.</param>
    /// <param name="Shadows">Shadow rendering mode.</param>
    public sealed record LightState(
        bool IsEnabled,
        LightType Type,
        Color Color,
        double Intensity,
        double Range,
        double OuterSpotAngle,
        double InnerSpotAngle,
        ShadowMode Shadows
    )
    {
        public LightState()
            : this(true, LightType.Point, Masonry.Color.White, 1, 10, 30, 0, ShadowMode.None) { }
    }

    /// <summary>Stable Animator state reconstructed by a snapshot.</summary>
    /// <param name="State">Animator state name to play.</param>
    /// <param name="Layer">Nonnegative Animator layer index.</param>
    /// <param name="NormalizedStartTime">Normalized starting time in [0, 1].</param>
    /// <param name="BoolParameters">Persistent boolean parameters by name.</param>
    /// <param name="IntParameters">Persistent signed integer parameters by name.</param>
    /// <param name="FloatParameters">Persistent finite floating parameters by name.</param>
    /// <param name="Speed">Nonnegative Animator playback speed.</param>
    public sealed record AnimatorState(
        string State,
        uint Layer,
        double NormalizedStartTime,
        IReadOnlyDictionary<string, bool> BoolParameters,
        IReadOnlyDictionary<string, int> IntParameters,
        IReadOnlyDictionary<string, double> FloatParameters,
        double Speed
    )
    {
        public AnimatorState(string state)
            : this(
                state,
                0,
                0,
                new Dictionary<string, bool>(),
                new Dictionary<string, int>(),
                new Dictionary<string, double>(),
                1
            ) { }
    }
}
